### PR TITLE
add plot method for eigenvalues

### LINKED ISSUE #159: Plot eigenvalues
Single-cell toolbox and mia do not currently have function for visualizing eigenvalues. Usually, that it done with screeplot that shows how the eigenvalues decreases.

1.

The function name could be `plotScree` or `plotEigenvalues`

2.

We could have methods for `SingleCellExperiment`, and `vector`.

SingleCellExperiment method gets the value from `reducedDim(x, dimred)`, from the correct attribute like functions `plotRDA()` and `plotLoadings()`. It creates a vector from the data with component names in `names()` and call the vector function.

Note that not all functions store the eigenvalues to same place / under same attribute name.

As usually eigenvalues are returned as a vector, we could have a method for vector also. That function could just create a data.frame from the vector. The data.frame could have the eigenvalues in one column and the names of vector in another. 

3.

Eigenvalues are commonly visualized with combination of barplot, point plot, line plot and labels. We could have them all but user could control them with `show.*`. These could be hidden parameters to make the methods less complex.

4.

Moreover, we could have also a cumulative explained variance plot that shows how much variance is explained by the first i component. That could also be a hidden option. Maybe `cumulative = FALSE`. In theory, that just needs one extra step that calculates the cumulative proportion.

image

### PR DESCRIPTION
**Pull request by 0xMuluh**
issue: #159

### CODE REVIEW

**Comment by TuomasBorman** [comment]
Commenting wirh phone so little bit hard, but:

1.

Eventhough SCE method, in description of x say that the obkect must be TreeSE (not SCE). Some users might not knoe that TreeSE is actually SCE and this might be confusing

2.

Structure the function so that. One internal functiona prepares the data and another internal function plots the data (.scree_plotter).

3.

Put these show.* parameters as a hidden parameters of .scree_plotter

**Comment by 0xMuluh** [comment]
i have added update to consider actual variance explained(`VarExplained`) by each PCi in `runPCA` as it does not return eigenvalues. What other `reducedDim` are to be supported?

**Comment by antagomir** [comment]
You could also support MDS (aka PCoA). For other ordinations this is not well defined I think.

File: R/plotScree.R [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        # Call the vector method
+        plotScree(as.numeric(eig), names(eig), cumulative = cumulative, ...)
+    }
+)
+
+#' @rdname plotScree
+#' @export
+setMethod("plotScree", signature = c(x = "vector"),
+    function(x, names = NULL, cumulative = FALSE, ...) {
+        # Ensure 'x' is numeric
```
Comment (TuomasBorman) [inline R/plotScree.R]: Wondering if this is better approach, so we can remove `names` argument

```
eigenvalues <- c(2, 3, 2)
names(eigenvalues) <- c("PC1", "PC2", "PC3")
plotScree(eigenvalues)
```

File: R/plotScree.R [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        if( any(ind) ){
+            # Add explained variance
+            eig <- attributes(reduced_dim)[ind][[1]]
+        } else{
+            stop("No eigenvalues found in the specified reducedDim.", 
+                 call. = FALSE)
+        }
+      
+        # Call the vector method
+        plotScree(as.numeric(eig), names(eig), cumulative = cumulative, ...)
```
Comment (TuomasBorman) [inline R/plotScree.R]: End functions with return()

File: R/plotScree.R [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+setGeneric("plotScree", signature = c("x"),
+    function(x, ...) standardGeneric("plotScree"))
+
+#' @rdname plotScree
+#' @export
+setMethod("plotScree", signature = c(x = "SingleCellExperiment"),
+    function(x, dimred, cumulative = FALSE, ...) {
+        # Check if dimred exists
+        if (!dimred %in% reducedDimNames(x)) {
+            stop("'dimred' must specify a valid reducedDim.", call. = FALSE)
```
Comment (TuomasBorman) [inline R/plotScree.R]: We could also support index (that could be supported but not described in the documentation --> simpler when less options)

File: R/plotScree.R [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        # Extract eigenvalues
+        # Check if data is available
+        ind <- names(attributes(reduced_dim)) %in% c("eig", "varExplained")
+        if( any(ind) ){
+            # Add explained variance
+            eig <- attributes(reduced_dim)[ind][[1]]
+        } else{
+            stop("No eigenvalues found in the specified reducedDim.", 
+                 call. = FALSE)
+        }
```
Comment (TuomasBorman) [inline R/plotScree.R]:
```
if( !any(ind) ){
   stop()
}
eig <- attributes(reduced_dim)[ind][[1]]
```

File: R/plotScree.R [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            vjust = -0.5)
+    }
+    
+    # Customize appearance
+    p <- p + theme_minimal() +
+        labs(x = "Component", 
+             y = if (cumulative) "Cumulative Proportion of Variance" 
+             else "Eigenvalue",
+             title = if (cumulative) "Cumulative Explained Variance" 
+             else "Scree Plot")
```
Comment (TuomasBorman) [inline R/plotScree.R]: Do not add title

File: R/plotScree.R [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    }
+    if (show.labels) {
+        p <- p + geom_text(aes(label = round(if (cumulative) 
+            CumulativeProportion else Eigenvalue, 2)), 
+            vjust = -0.5)
+    }
+    
+    # Customize appearance
+    p <- p + theme_minimal() +
+        labs(x = "Component",
```
Comment (TuomasBorman) [inline R/plotScree.R]: X axis title could be PC so it is more general

File: R/plotScree.R [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        .scree_plotter(plot_data, cumulative = cumulative, ...)
+    }
+)
+
+.prepare_data <- function(x, names = NULL, cumulative = FALSE) {
+    df <- data.frame(
+        Component = if (!is.null(names)) names else seq_along(x),
+        Eigenvalue = x
+    )
+
```
Comment (TuomasBorman) [inline R/plotScree.R]: Remove extra line breaks, this is so short function so they do not improve readability

File: R/plotScree.R [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+.prepare_data <- function(x, names = NULL, cumulative = FALSE) {
+    df <- data.frame(
+        Component = if (!is.null(names)) names else seq_along(x),
+        Eigenvalue = x
+    )
+    
+    # Calculate cumulative proportion if needed
+    if (cumulative) {
+        df$CumulativeProportion <- cumsum(df$Eigenvalue) / sum(df$Eigenvalue)
```
Comment (TuomasBorman) [inline R/plotScree.R]: What if there are NAs?

**Comment by TuomasBorman** [comment]
> You could also support MDS (aka PCoA). For other ordinations this is not well defined I think.

Also RDA is based on eigenvalues-eigenvectors

File: R/plotScree.R (hunk 8 above)
Comment (0xMuluh) [reply R/plotScree.R]: issued warning and removed NAs.

File: R/plotScree.R (hunk 1 above)
Comment (0xMuluh) [reply R/plotScree.R]: ACK

File: R/plotScree.R (hunk 3 above)
Comment (0xMuluh) [reply R/plotScree.R]: I'm not sure I understand this. Do you mean subsetting the vector based on indices for example `indices = c(4,5)`?

Comment (TuomasBorman) [reply R/plotScree.R]: Sorry about unclear message, I meant `reducedDim(tse, 1)`, but I think you did it correctly

**Pull request merged**